### DISCUSSION TITLE
Plugins

### DISCUSSION THREAD

**Discussion by slook** [nicotine-plus/nicotine-plus#1540] [General]
This topic is for the plugin system, to enable other developers to add functionality with community developed modules.

- This is not a conversation thread, and as such, please keep replies in-line and on-topic.

**Comment by slook** [comment]
**Methods and properties available for developers**

What functions are possible?

Is some documentation of the development of plugins needed?

**Reply by slook** [reply]
Perhaps the documentation could be integrated as a heavily commented package containing a "Hello World" plugin?

**Reply by slook** [reply]
The most common functions seem to be related around chat and music players.

I would like to add TreeView columns, but this is not currently possible, because no hook is provided.

**Comment by slook** [comment]
**Security**

This needs thinking more when adding native plugins into N+ core.

**Reply by slook** [reply]
Plugin developers should not need to modify core files to add or updates plugins.

**Reply by slook** [reply]
URL's should be removed from the core. We should consider making our own hard-coded web functions into plugins, or at the very least, a configuration file.

**Reply by slook** [reply]
**Approved vs Third-party**: Shall it be wise for us to maintain a list of approved plugins? And warn end-users if unverified plugins are about to be installed. Or will this be too intrusive?

**Reply by mathiascode** [reply]
I don't personally want to pick up yet another responsibility of reviewing third-party plugins.

**Reply by slook** [reply]
Then in that case, all non-standard plugins (Author=not "Nicotine+ Team") can be more clearly identified as such.

Preferences > Plugins > "Enable Plugins" checkbox: suggest change to two separate checkboxes...

- "Enable Built-in Plugins"
- "Enable Third-party Plugins" _with ToolTip "Caution: installing third-party plugins may cause unexpected results"_
 
_(deseslect disables/greyout individual list items)_

**Reply by Nachtalb** [reply]
Instead of separate checkboxes, you could just open a popup warning for enabling a third-party plugin for the first time or put a little warning in the metadata section of the plugin with something like: 
> ⚠️ This is not a built-in plugin. Use at own risk.

**Reply by mathiascode** [reply]
> Instead of separate checkboxes, you could just open a popup warning for enabling a third-party plugin for the first time or put a little warning in the metadata section of the plugin with something like:
> 
> > ⚠️ This is not a built-in plugin. Use at own risk.

This is now done in master.

**Comment by Nachtalb** [comment]
# How to improve plugin architecture 

**Any ideas are welcome!**

Main questions: 
- Is the current plugin architecture future proof?
- Should the architecture be changed? Here are some structures:

	- Provider / Adapter / Interface architecture with [zope.interface](https://pypi.org/project/zope.interface/) (and maybe more). An example here would be [Plone CMS](https://plone.org/), an incredible powerful CMS build on top of the [Zope suite](https://zope.org/).
	- Exposing a greater portion of N+ as safe to use API for plugins.
	- Expanding the current plugin system with more high-level functionality (eg. easily providing menu options or status bar buttons etc) and providing more hooks like eg. triggering a hook before a command is run in order to override the outcome like with [notification events](https://github.com/nicotine-plus/nicotine-plus/blob/13d8786eecb136c4ee1f1172864ed58d1dd11cd9/pynicotine/chatrooms.py#L128-L139).

**Reply by mathiascode** [reply]
In order to keep Nicotine+ portable across the vast number of systems and distros we currently support, we have to stick with functionality provided by the Python standard libraries. While it's too late to make any significant changes to the plugin system, I agree that we should add more "safe" functions and hooks that plugin developers can use.

**Comment by Nachtalb** [comment]
# Things that need addressing now and in the future

## important

- [FIXED] Crash: An error in a plugin must not should the entire N+ instance
- Multi-Module Plugins: At the moment it is not possible to load relative modules without path injection: https://github.com/Nachtalb/more-upload-stats/blob/40e7c523e4b3edd958a158d062edfbd243b1c4a9/__init__.py 

## should at least

- Run Commands: Being able to run commands other than `/me`. In other words, cutting down [this behemoth](https://github.com/nicotine-plus/nicotine-plus/blob/13d8786eecb136c4ee1f1172864ed58d1dd11cd9/pynicotine/gtkgui/widgets/textentry.py#L215) into a command handler or something alike [DONE]
- [DONE] Expanding command help window with plugin commands and providing auto-complete functionality

## quirks 

- [FIXED] Missing modules:  During my plugin development I noticed various quirks like built-in python modules not being available like `dataclasses` and `html` even tho the python version reported by the plugin through `sys` is `3.9.4`.

**Reply by slook** [reply]
> Crash: An error in a plugin must not crash the entire N+ instance

Perhaps @Nachtalb is onto something here, this is a topic that may require some attention @mathiascode ?

Also see https://github.com/nicotine-plus/nicotine-plus/discussions/1533#discussioncomment-1279341 and https://github.com/nicotine-plus/nicotine-plus/discussions/1533#discussioncomment-1280477

> python version reported by the plugin

Note backward compatible OS requires support for Python 3.5 until June 2022.

**Reply by mathiascode** [reply]
> Crash: An error in a plugin must not should the entire N+ instance

Can this be resolved while retaining the flexibility plugins have at the moment? There are potentially endless ways to crash/freeze Nicotine+ with a bit of Python.

**Reply by slook** [reply]
Indeed. That's why such things need to be actively avoided in the core code. Anything could happen otherwise.

**Reply by mathiascode** [reply]
> Missing modules:  During my plugin development I noticed various quirks like built-in python modules not being available like `dataclasses` and `html` even tho the python version reported by the plugin through `sys` is `3.9.4`.

This is a Windows/macOS issue, since [PyInstaller](https://www.pyinstaller.org/) only bundles the Python modules required by Nicotine+.

**Reply by Nachtalb** [reply]
> This is a Windows/macOS issue, since [PyInstaller](https://www.pyinstaller.org/) only bundles the Python modules required by Nicotine+.

One possible solution: https://stackoverflow.com/a/62738273/5699307

**Reply by Nachtalb** [reply]
> Can this be resolved while retaining the flexibility plugins have at the moment? There are potentially endless ways to crash/freeze Nicotine+ with a bit of Python.

I'll have a look once I have an N+ source build up and running. NGL tho I am doing some hacky things in my plugins and I could resolve it in my plugins myself. It mainly happened when an error in a spawned thread occurs (I could catch that myself). But I am sure there is some way of preventing a total crash from N+'s side as well. But it is true that with python you can of course hack your way all around and cause issues....

> We are all consenting adults
> ~ https://python-guide-chinese.readthedocs.io/zh_CN/latest/writing/style.html#we-are-all-consenting-adults

**Reply by Nachtalb** [reply]
> Note backward compatible OS requires support for Python 3.5 until June 2022.

On that note. Where does that os compatibility come from? According to python 3.5 is EOL and so is 3.6 soon (end of 2021)
https://devguide.python.org/devcycle/#end-of-life-branches

I don't think it should be supported much longer than the official support. Thus IMHO 3.5 should be abandoned ASAP :/

**Reply by slook** [reply]
> Where does that os compatibility come from?

Debian 9 Stretch (LTS), at present =>3.6 does not run properly on some major debian based distro's. See https://github.com/nicotine-plus/nicotine-plus/issues/1488

https://wiki.debian.org/Python#Debian_Python_Policy_for_Python_developers

**Reply by mathiascode** [reply]
> One possible solution: https://stackoverflow.com/a/62738273/5699307

This looks fine, and it seems like the installer/package size only increases by a few MBs, which is reasonable.

**Reply by mathiascode** [reply]
> Crash: An error in a plugin must not should the entire N+ instance

Can you give some examples of situations you've encountered?

**Reply by mathiascode** [reply]
Can you check if all the modules you need are included in the latest unstable build?

**Reply by Nachtalb** [reply]
- Malformed PLUGININFO (eg. missing `Version`)
- An error occurse in a spawned thread: Please take a look at #1546
- ...

**Reply by mathiascode** [reply]
[DONE] Once https://github.com/nicotine-plus/nicotine-plus/pull/1581 is merged, default chat commands will be moved to a hidden "Core Commands" plugin.

**Comment by mathiascode** [comment]
*WONTFIX* - **Deprecate plugin notification system**

Since we now use the new event system for communication in Nicotine+ 3.3.0, we could start using it in plugins as well. Plugin authors would get any newly added events for free.

https://github.com/nicotine-plus/nicotine-plus/blob/master/pynicotine/events.py

**Reply by slook** [reply]
The ability to use the event emitter in the plugins would be highly flexible, but since that is quite far removed from standard Pythonic programming that plugin developers would expect, I think most would probably prefer for selected events to be abstracted into standard override methods.

Maybe there could be a way for the pluginsystem to do this sort of method name abstraction automatically, so that we don't have to maintain the overrides each time a new event is added. On the other hand, it would be difficult to find out what arguments are used.

**Reply by mathiascode** [reply]
I'm going to mark this as a wontfix. There's some value in having simpler plugin notifications for common use cases, and we don't need to break every plugin out there.

**Reply by mathiascode** [reply]
Plugin authors can still use the lower-level event system at their own risk, if they want more power.

**Comment by deafwired** [comment]
# html parser
Correct me if im wrong, but the html module is part of the standard python library. When i import it my plugin works just fine, but when i try ```from html.parser import HTMLParser``` it doesn't work. I would love to use this in my plugins, and I saw that it is an issue around pyinstaller, but cant there be a way to add basic modules like this? (I am using the current stable build of nicotine+)

**Reply by mathiascode** [reply]
What's the error message? Everything included in the Python standard library should be available in the Windows and macOS builds of Nicotine+ nowadays.

**Reply by deafwired** [reply]
With a simple plugin such as this
```
from pynicotine.pluginsystem import BasePlugin
from html import HTMLParser

class Plugin(BasePlugin):
    def __init__(self):
        super().__init__()
        self.commands = {
            "callback": self.test,
            "description": "Test plugin",
            "usage": ["test"],
            "usage_private_chat": ["test"]
        }
    def test(self):
        self.log("Test plugin loaded")
    def log(self, *msg):
        super().log(*msg)
```
I get
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/0c5640a9-4422-45b7-bccf-1bad3aefe243)
My version
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/02ef5325-6277-4307-b204-2021505051a4)

**Reply by slook** [reply]
_See above https://github.com/nicotine-plus/nicotine-plus/discussions/1540#discussioncomment-1280612_:
> [FIXED] Missing modules: During my plugin development I noticed various quirks like built-in python modules not being available like `dataclasses` and `html` even tho the python version reported by the plugin through sys is 3.9.4.


Do you get the same error on 3.3.0.dev5?

Nicotine+ 3.3.0.dev5: https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md

**Reply by deafwired** [reply]
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/7f8b87c1-b06c-4b6d-b798-b681023585c0)
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/c0f6b54a-4065-408b-98a8-1b5effa76649)
Just installed it and get the same error

**Reply by Inso-m-niaC** [reply]
You need the html5lib package to be able to import the parser.

**Reply by mathiascode** [reply]
@Inso-m-niaC Ideally this shouldn't be necessary, but cx_Freeze isn't including a handful of standard Python modules for some reason. Not sure why yet.

**Reply by Inso-m-niaC** [reply]
The module is there. But, it is missing parser.py. The other two files are present (init and entities).

**Reply by deafwired** [reply]
@mathiascode If you are able to create a solution, would something like the HTMLParser be included?

**Reply by mathiascode** [reply]
> The module is there. But, it is missing parser.py. The other two files are present (init and entities).

Python technically treats html.parser as a separate (sub)module. I assume nothing else in the standard library uses it, which results in cx_Freeze excluding it from our builds.

@deafwired https://github.com/nicotine-plus/nicotine-plus/commit/178905907d30ba65611311255179a40d3cdc97be should fix the issue. Try the latest development build in about 5-10 minutes: https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md

**Reply by deafwired** [reply]
I installed the unstable version (3.3.0.dev5) and tried importing HTMLParser from html and get the same error, but when i look in library.zip, the parser.pyc file is there.
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/88d9e533-b365-4ecc-9860-045291b24500)
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/73e1e98b-cc43-4041-b460-751b08cd3a74)

**Reply by mathiascode** [reply]
Use `from html.parser import HTMLParser` instead of `from html import HTMLParser`

**Reply by deafwired** [reply]
Sorry for my stupid mistake, it does seem to load now thank you!

**Reply by slook** [reply]
You might as well stick with 3.3.x for developing your plugins, to take advantage of the new command system and future improvements to the core pluginsystem module etc. The 3.2.x branch is not being developed anymore.

If there's anything you need, please let us know.

Please note `"usage"` -> `"parameters"`

**Reply by deafwired** [reply]
i'm sorry, what/where is usage -> parameters

**Reply by slook** [reply]
I meant the `commands` dict key has changed from "usage" (and "usage_private_chat" etc) in 3.3.0.dev1 to "parameters" (and "parameters_private_chat" etc) in 3.3.0.dev5. For example:

```
        self.commands = {
            "test": {
                "callback": self.test,
                "description": "Test plugin",
                "parameters": ["<test>"],
                "parameters_private_chat": ["[test]"]
            }
        }

```

`<...>` denotes a required positional argument.
`[...]` denotres an optional argument (such as in `parameters_private_chat` if "test" is a username that can already be known).

See the [core_commands plugin](https://github.com/nicotine-plus/nicotine-plus/blob/cadc221710597a3a1d55fb1639b067938fb718a8/pynicotine/plugins/core_commands/__init__.py#L41-L265) or [port_checker plugin](https://github.com/nicotine-plus/nicotine-plus/blob/cadc221710597a3a1d55fb1639b067938fb718a8/pynicotine/plugins/examplars/port_checker/__init__.py#L46-L53) for more examples.

The plugin commands interpreter is in the process of being ratified, so it is subject to being amended again before the stable release of 3.3.0 upon the feedback of plugin developers.
